modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java

```diff
@@ -2749,107 +2822,105 @@ private double getPrefLength() {
-
-        // "p" represents the most recent value for position. This is always
-        // based on the edge between two cells, except at the very end of the
-        // algorithm where it is added to the computed "p" offset for the final
-        // value of Position.
-        double p = cellPercent * cellIndex;
-
-        // Loop over the cells one at a time until either we reach the end of
-        // the cells, or we find that the "n" will fall within the cell we're on
-        while (n > remaining && ((forward && cellIndex < cellCount - 1) || (! forward && cellIndex > 0))) {
-            if (forward) cellIndex++; else cellIndex--;
-            n -= remaining;
-            cellSize = getCellLength(cellIndex);
-            start = computeOffsetForCell(cellIndex);
-            end = cellSize + computeOffsetForCell(cellIndex + 1);
-            remaining = end - start;
-            p = cellPercent * cellIndex;
-        }
-
-        // if remaining is < n, then we must have hit an end, so as a
-        // fast path, we can just set position to 1.0 or 0.0 and return
-        // because we know we hit the end
-        if (n > remaining) {
-            setPosition(forward ? 1.0f : 0.0f);
-        } else if (forward) {
-            double rate = cellPercent / Math.abs(end - start);
-            setPosition(p + (rate * n));
-        } else {
-            double rate = cellPercent / Math.abs(end - start);
-            setPosition((p + cellPercent) - (rate * n));
+    private double adjustByPixelAmount(double numPixels) {
+        if (numPixels == 0) return 0;
+        // When we're at the top already, we can't move back further, unless we
+        // want to allow for gravity-alike effects.
+        if ((absoluteOffset <= 0) && (numPixels < 0)) return 0;
+
+        // start with applying the requested modification
+        double origAbsoluteOffset = this.absoluteOffset;
+        this.absoluteOffset = Math.max(0.d, this.absoluteOffset + numPixels);
+        double newPosition = Math.min(1.0d, absoluteOffset/(estimatedSize - viewportLength));
+        // estimatedSize changes may result in opposite effect on position
+        // in that case, modify current position 1% in the requested direction
+        if ((numPixels > 0) && (newPosition < getPosition())) {
+            newPosition = getPosition()*1.01;
         }
+        if ((numPixels < 0) && (newPosition > getPosition())) {
+            newPosition = getPosition()*.99;
+        }
+
+        // once at 95% of the total estimated size, we want a correct size, not
+        // an estimated size anymore.
+        if (newPosition > .95) {
+            int cci = computeCurrentIndex();
+            while (cci < getCellCount()) {
+                getOrCreateCellSize(cci); cci++;
+            }
+            recalculateEstimatedSize();
+        }
+
+        // if we are at or beyond the edge, correct the absolteOffset
+        if (newPosition >= 1.d) {
```

Comment:
Do you need to also check for `newPosition < 0`  if you are scrolling backwards?

Reply:
That can not happen. The only reason we can have `newPosition > 1.d` is when the correction applied to scrolling in the wrong direction would bring the position beyond 1:
`newPosition = getPosition()*1.01`

The inverse check for this might reduce the position (by `.99`) but it will never render it negative.

Note that there is a documentation placeholder in case we want to allow for gravity-alike scrolling 
```  
      // When we're at the top already, we can't move back further, unless we
      // want to allow for gravity-alike effects.
```
In that case, we might want to use negative values for the position, but that seems a follow-up issue to me (if we want to do that at all, that is)